### PR TITLE
Separate logging and reqid options

### PR DESCRIPTION
**Pull request by cemremengu**
Addresses #1035

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/reqIdGenFactory.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = function (requestIdHeader) {
```
Comment (allevo) [inline lib/reqIdGenFactory.js]: Could you add `'use strict'` on the top of the file?

File: docs/Server.md [hunk 2]
```diff
@@ -154,6 +154,21 @@ The header name used to know the request id. See [the request id](https://github
 
 + Default: `'request-id'`
 
+<a name="factory-gen-request-id"></a>
+### `genReqId`
+
+Function for generating the request id. It will receive the incoming request as a parameter.
```
Comment (allevo) [inline docs/Server.md]: `Function for generating the request id`=> `Function for generating the request id if the request id header is not in the request`?

**Comment by cemremengu** [comment]
You may want to label this PR as `semver-major`

File: docs/Server.md (hunk 2 above)
Comment (allevo) [reply docs/Server.md]: Sorry, my bad. This is not true. If the request id generator is provided, the id is not read from request header. Could you undo this change?

Comment (cemremengu) [reply docs/Server.md]: actually it is `req.headers[requestIdHeader] || (nextReqId = (nextReqId + 1) & maxInt)` in the default generator function implementation so I guess it is true?

Comment (cemremengu) [reply docs/Server.md]: But if you override it, then yes the id is not read from request header unless you implement it that way

Comment (cemremengu) [reply docs/Server.md]: Undoing the change and adding `+ Default: `value of 'request-id' if provided or monotonically increasing integers`

File: fastify.js [hunk 3]
```diff
@@ -85,15 +86,15 @@ function build (options) {
...
 
   const requestIdHeader = options.requestIdHeader || 'request-id'
+  const genReqId = options.genReqId || reqIdGenFactory(requestIdHeader)
 
   fastify.printRoutes = router.prettyPrint.bind(router)
 
   const setupResponseListeners = Reply.setupResponseListeners
+
   // logger utils
-  const customGenReqId = options.logger ? options.logger.genReqId : null
```
Comment (mcollina) [inline fastify.js]: Can you throw if genReqId is specified under logger? Or move it to the main option object and emit a warning (process.emitWarning)?

Comment (cemremengu) [reply fastify.js]: #1166

**Pull request merged**